File: stock-market-service/src/main/java/rs/edu/raf/service/StockService.java
package rs.edu.raf.service;

import rs.edu.raf.dto.StockDTO;
import rs.edu.raf.dto.StockHistoricalDataDTO;

public interface StockService {
    /**
     * Retrieves stock with specified ticker
     * @param ticker the ticker
     * @return {@link StockDTO} object representing stock
     */
    StockDTO getStockByTicker(String ticker);

    /**
     * Retrieves daily historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     * @return {@link StockHistoricalDataDTO} object representing historical data
     */
    StockHistoricalDataDTO getDailyHistoricalDataForTicker(String ticker);

    /**
     * Retrieves weekly historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     * @return {@link StockHistoricalDataDTO} object representing historical data
     */
    StockHistoricalDataDTO getWeeklyHistoricalDataForTicker(String ticker);

    /**
     * Retrieves monthly historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     * @return {@link StockHistoricalDataDTO} object representing historical data
     */
    StockHistoricalDataDTO getMonthlyHistoricalDataForTicker(String ticker);

    /**
     * Updates all stocks with data from external API
     */
    void updateAllStocks();

    /**
     * Fetch all stocks from external API and cache in redis
     */
    void fetchAllStocks();

    /**
     * Fetch stock with specified ticker
     * @param ticker the ticker of the stock
     */
    void fetchStockWithTicker(String ticker);

    /**
     * Fetch daily historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     */
    void fetchDailyHistoricalData(String ticker);

    /**
     * Fetch weekly historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     */
    void fetchWeeklyHistoricalData(String ticker);

    /**
     * Fetch daily historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     */
    void fetchMonthlyHistoricalData(String ticker);
}

File: stock-market-service-kotlin/src/main/kotlin/org/example/stockmarketservicekotlin/service/StockService.kt
package org.example.stockmarketservicekotlin.service

import org.example.stockmarketservicekotlin.dto.StockDTO
import org.example.stockmarketservicekotlin.dto.StockHistoricalDataDTO

interface StockService {
    /**
     * Retrieves stock with specified ticker
     * @param ticker the ticker
     * @return [StockDTO] object representing stock
     */
    fun getStockByTicker(ticker: String): StockDTO

    /**
     * Retrieves daily historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     * @return [StockHistoricalDataDTO] object representing historical data
     */
    fun getDailyHistoricalDataForTicker(ticker: String): StockHistoricalDataDTO

    /**
     * Retrieves weekly historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     * @return [StockHistoricalDataDTO] object representing historical data
     */
    fun getWeeklyHistoricalDataForTicker(ticker: String): StockHistoricalDataDTO

    /**
     * Retrieves monthly historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     * @return [StockHistoricalDataDTO] object representing historical data
     */
    fun getMonthlyHistoricalDataForTicker(ticker: String): StockHistoricalDataDTO

    /**
     * Updates all stocks with data from external API
     */
    fun updateAllStocks()

    /**
     * Fetch all stocks from external API and cache in redis
     */
    fun fetchAllStocks()

    /**
     * Fetch stock with specified ticker
     * @param ticker the ticker of the stock
     */
    fun fetchStockWithTicker(ticker: String)

    /**
     * Fetch daily historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     */
    fun fetchDailyHistoricalData(ticker: String)

    /**
     * Fetch weekly historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     */
    fun fetchWeeklyHistoricalData(ticker: String)

    /**
     * Fetch daily historical data for stock with specified ticker
     * @param ticker the ticker of the stock
     */
    fun fetchMonthlyHistoricalData(ticker: String)
}